### PR TITLE
feat: Create endpoint in Charges Web API that serves charge history data

### PR DESCRIPTION
**Pull request by prtandrup**
## Description

With this PR a new Charges Web API endpoint is created that upon request can provide Charge History data (Charge Information).

- [x] `ChargeHistoryController` and test added
- [x] Seeded Charge History test data added, so it may be used to fulfill the ACs, verified through the `ChargeHistoryQueryServiceTests`
- [x] `ChargeHistoryQueryService` must be able to deliver the "actual timeline" based on the AtDateTime provided in the search criteria
- [x] Deliver correct EndDateTime when charge has been stopped

## References

* #1846

### CODE REVIEW

File: source/Contracts/ChargeHistory/ChargeHistorySearchCriteriaV1Dto.cs [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+using Energinet.DataHub.Charges.Contracts.Charge;
+
+// ReSharper disable once CheckNamespace - Type is shared so namespace is not determined by project structure/namespace
+namespace Energinet.DataHub.Charges.Contracts.ChargeHistory
+{
+    public record ChargeHistorySearchCriteriaV1Dto(
+        string ChargeId, // SenderProvidedChargeId
+        ChargeType ChargeType,
+        string ChargeOwner, // Market Participant GLN no.
+        DateTimeOffset AtDateTimeUtc); // View the Charge history timeline from this point in (date) time.
```
[the remark this answers is not in the document]
Comment (prtandrup) [reply source/Contracts/ChargeHistory/ChargeHistorySearchCriteriaV1Dto.cs:25]: With the date range picker you'll not be able to query history data at a certain point in time, so in my opinion the GUI needs to be changed to a single datetime input. Currently, I'm not seeing a need for the date range picker in that use case/context.

File: source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.QueryApi/QueryServices/ChargeHistoryQueryService.cs [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+    public class ChargeHistoryQueryService : IChargeHistoryQueryService
+    {
+        private readonly IData _data;
+
+        public ChargeHistoryQueryService(IData data)
+        {
+            _data = data;
+        }
+
+        public async Task<IList<ChargeHistoryV1Dto>> GetAsync(ChargeHistorySearchCriteriaV1Dto searchCriteria)
```
[the remark this answers is not in the document]
Comment (prtandrup) [reply source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.QueryApi/QueryServices/ChargeHistoryQueryService.cs]: done in https://github.com/Energinet-DataHub/geh-charges/pull/1877/commits/e1479c6f9b53892bde7acdded912840c513897c3

File: source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.QueryApi/QueryServices/IChargeHistoryQueryService.cs [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+/// <summary>
+/// Service for querying charge history
+/// </summary>
+public interface IChargeHistoryQueryService
+{
+    /// <summary>
+    /// Get charge history for a single charge
+    /// </summary>
+    /// <returns>A <see cref="ChargeHistoryV1Dto"/></returns>
+    Task<IList<ChargeHistoryV1Dto>> GetAsync(ChargeHistorySearchCriteriaV1Dto searchCriteria);
```
[the remark this answers is not in the document]
Comment (prtandrup) [reply source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.QueryApi/QueryServices/IChargeHistoryQueryService.cs]: done in https://github.com/Energinet-DataHub/geh-charges/pull/1877/commits/e1479c6f9b53892bde7acdded912840c513897c3

File: source/Contracts/ChargeHistory/ChargeHistorySearchCriteriaV1Dto.cs [hunk 4]
```diff
@@ -0,0 +1,26 @@
...
+using System;
+using Energinet.DataHub.Charges.Contracts.Charge;
+
+// ReSharper disable once CheckNamespace - Type is shared so namespace is not determined by project structure/namespace
+namespace Energinet.DataHub.Charges.Contracts.ChargeHistory
+{
+    public record ChargeHistorySearchCriteriaV1Dto(
+        string ChargeId, // SenderProvidedChargeId
+        ChargeType ChargeType,
+        string ChargeOwner, // Market Participant GLN no.
```
Comment (HenrikSommer) [inline source/Contracts/ChargeHistory/ChargeHistorySearchCriteriaV1Dto.cs:24]: Why not just `ChargeId` and then let the backend load the Charge to get the combined key, like it is done in `ChargeMessagesSearchCriteriaV1Dto` and `ChargeMessageQueryService`?

File: source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.QueryApi/QueryServices/ChargeHistoryQueryService.cs [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+        }
+
+        private static IList<ChargeHistoryV1Dto> MapToChargeHistoryV1Dtos(IList<ChargeHistory> chargeHistories)
+        {
+            var lastIndex = chargeHistories.IndexOf(chargeHistories.Last());
+
+            return chargeHistories
+                .Select((c, i) => new ChargeHistoryV1Dto(
+                    DateTime.SpecifyKind(c.StartDateTime, DateTimeKind.Utc),
+                    i == lastIndex ? DetermineEndDateTime(c) : DateTime.SpecifyKind(chargeHistories[i + 1].StartDateTime, DateTimeKind.Utc),
```
Comment (HenrikSommer) [inline source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.QueryApi/QueryServices/ChargeHistoryQueryService.cs:60]: Interesting code :-)
I had to read it twice or more to get it. Maybe we could put it in a method with a descriptive name?

Comment (prtandrup) [reply source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.QueryApi/QueryServices/ChargeHistoryQueryService.cs]: Please have a look at the refactored code, better? https://github.com/Energinet-DataHub/geh-charges/pull/1877/commits/cf0fcd743210824317a3e9eb65c4f19bbfc3257c

File: source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.IntegrationTests/IntegrationTests/WebApi/QueryServices/ChargeHistoryQueryServiceTests.cs [hunk 6]
```diff
@@ -0,0 +1,213 @@
...
+            using var assertionScope = new AssertionScope();
+
+            actual[0].StartDateTime.Should().Be(TestData.ChargeHistory.TariffB.FirstStartDateTime);
+            actual[0].EndDateTime.Should().Be(TestData.ChargeHistory.TariffB.SecondStartDateTime);
+            actual[1].StartDateTime.Should().Be(TestData.ChargeHistory.TariffB.SecondStartDateTime);
+            actual[1].EndDateTime.Should().Be(TestData.ChargeHistory.TariffB.SecondStartDateTime);
+        }
+
+        [Fact]
+        public async Task SearchAsync_WhenChargeDoesntExist_ThrowsArgumentException()
```
Comment (HenrikSommer) [inline source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.IntegrationTests/IntegrationTests/WebApi/QueryServices/ChargeHistoryQueryServiceTests.cs:192]:
```suggestion
        public async Task SearchAsync_WhenChargeDoesNotExist_ThrowsArgumentException()
```

File: source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.ApplyDBMigrationsApp/Scripts/Test/202201261750 Add test data.sql [hunk 7]
```diff
@@ -87,19 +90,23 @@ INSERT INTO charges.charge VALUES (@chargeId5, 'TestTaxTar', 3, @systemOperatorI
...
-    2, /* Resolution, P1D */
-    1, /* TransparentInvoicing, yes */
-    '2014-12-31 23:00', /* StartDateTime */
+    'EA-001a', -- ChargeId
+    3, -- [Type], Tariff
+    @chargeOwnerId, -- Owner
+    1, -- TaxIndicator, yes
+    2, -- Resolution, PT1H
+    1, -- TransparentInvoicing, yes
+    '2014-12-31 23:00', --StartDateTime
```
Comment (HenrikSommer) [inline source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.ApplyDBMigrationsApp/Scripts/Test/202201261750 Add test data.sql:99]:
```suggestion
    '2014-12-31 23:00', -- StartDateTime
```

File: source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.QueryApi/QueryServices/ChargeHistoryQueryService.cs [hunk 8]
```diff
@@ -0,0 +1,98 @@
...
+            _data = data;
+        }
+
+        public async Task<IList<ChargeHistoryV1Dto>> SearchAsync(ChargeHistorySearchCriteriaV1Dto searchCriteria)
+        {
+            var charge = await _data.Charges
+                .SingleOrDefaultAsync(c => c.Id == searchCriteria.ChargeId)
+                .ConfigureAwait(false);
+
+            if (charge == null) throw new ArgumentException("Charge not found");
```
Comment (HenrikSommer) [inline source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.QueryApi/QueryServices/ChargeHistoryQueryService.cs:41]: Technically not an argument exception.

**Review by HenrikSommer — APPROVED** [review]

File: source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.QueryApi/QueryServices/ChargeHistoryQueryService.cs [hunk 9]
```diff
@@ -0,0 +1,112 @@
...
+        {
+            var lastIndex = chargeHistories.IndexOf(chargeHistories.Last());
+            var isChargeHistoryTheMostRecent = index == lastIndex;
+
+            return isChargeHistoryTheMostRecent
+                ? UseChargeStopDateOrNull(chargeHistory)
+                : UseNextChargeHistoryStartAsEndDateTimeForCurrentHistory(chargeHistories, index);
+        }
+
+        private static DateTimeOffset? UseNextChargeHistoryStartAsEndDateTimeForCurrentHistory(IList<ChargeHistory> chargeHistories, int index)
```
Comment (HenrikSommer) [inline source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.QueryApi/QueryServices/ChargeHistoryQueryService.cs:91]: ♥

**Review by HenrikSommer — APPROVED** [review]

**Pull request merged**